### PR TITLE
Fix #254

### LINKED ISSUE #254: Parsing crate `foo`: can't find mod `bar` with edition 2018 submodules.
```
$ rustup default beta
$ cargo new --lib foo
$ cd foo
$ echo "mod qux;" > src/lib.rs
$ echo "mod bar;" > src/qux.rs
$ mkdir src/qux
$ touch src/qux/bar.rs
$ cargo build
   Compiling foo v0.1.0 (/tmp/foo)                                                                                                                                       
    Finished dev [unoptimized + debuginfo] target(s) in 0.33s
$ cbindgen
WARN: Parsing crate `foo`: can't find mod bar`.
#include <cstdint>
#include <cstdlib>
```

### PR DESCRIPTION
**Pull request by vtavernier**
By checking the recursion depth in the crate we are currently parsing, we can decide on the right module directory to find submodules in. The added `mod_2018` test fails without the changes to `parser.rs`.

### CODE REVIEW

File: src/bindgen/parser.rs [hunk 1]
```diff
@@ -238,6 +238,10 @@ impl<'a> Parser<'a> {
     }
 
     fn parse_mod(&mut self, pkg: &PackageRef, mod_path: &FilePath) -> Result<(), Error> {
+        self.parse_mod_depth(pkg, mod_path, 0)
```
Comment (emilio) [inline src/bindgen/parser.rs:241]: Can we just pass zero on the relevant callers and keep `parse_mod` as the name for all functions?

File: src/bindgen/parser.rs [hunk 2]
```diff
@@ -265,16 +269,25 @@ impl<'a> Parser<'a> {
...
         };
 
-        let mod_dir = mod_path.parent().unwrap();
+        // Compute module directory according to Rust 2018 rules
+        let mod_dir_2018;
+
+        let mod_dir = if depth == 0 {
+            mod_path.parent().unwrap()
+        } else {
+            mod_dir_2018 = mod_path.parent().unwrap().join(mod_path.file_stem().unwrap());
```
Comment (emilio) [inline src/bindgen/parser.rs:278]: Does this break non-2018 users?

**Comment by emilio** [comment]
Thanks so much for working on this btw!

File: src/bindgen/parser.rs (hunk 1 above)
Comment (vtavernier) [reply src/bindgen/parser.rs]: I actually hesitated between the two options, fixed in 0d32d99

File: src/bindgen/parser.rs (hunk 2 above)
Comment (vtavernier) [reply src/bindgen/parser.rs:274]: It seems to break non-2018 unfortunately, I added a test for that in 0d89c9e and I'll investigate this in more detail.

Comment (vtavernier) [reply src/bindgen/parser.rs:274]: Should be fixed by 835229b

File: src/bindgen/parser.rs [hunk 3]
```diff
@@ -265,16 +265,25 @@ impl<'a> Parser<'a> {
             self.cache_src.get(&owned_mod_path).unwrap().clone()
         };
 
-        let mod_dir = mod_path.parent().unwrap();
+        // Compute module directory according to Rust 2018 rules
+        let mod_dir_2018;
 
-        self.process_mod(pkg, mod_dir, &mod_parsed)
+        let mod_dir = if depth == 0 || mod_path.ends_with("mod.rs") {
```
Comment (emilio) [inline src/bindgen/parser.rs:271]: Hmm, I still don't think this is quite sound is it? At the very least we should check the file name for full equality for `mod.rs`, not just whether the path ends with `mod.rs`

Comment (vtavernier) [reply src/bindgen/parser.rs:271]: Apparently `Path::ends_with` only checks entire path components so it is checking if the (entire) file name is `mod.rs`. This wouldn't be the case with a `&str`.

https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.ends_with

**Comment by emilio** [comment]
Sorry for the lag here, taking this patch for a spin now :)

**Pull request merged**